### ISSUE TITLE
classifier train seems to be broken (for cifar labeled data)

### ISSUE THREAD

**Issue by whentheworldfallsapart** [pjreddie/darknet#1262]
So their seems to be a weird issue with classifier training, if darknet net is built with cudnn (cuda=1 in the Makefile doesnt matter) then the loss will go down with training but when predict or valid is run the model always predicts the same classes with the same confidences. 
I have tried running the cifar example https://pjreddie.com/darknet/train-cifar/ on two different machines

Centos 7.4 with Cuda 9.0 and Cudnn 7 
Centos 7.2 with Cuda 8 and Cudnn 6 

both observed the same issue, strangely enough if i build with cudnn and change the 2nd from the last conv layer of the cifar_small cfg from filters=64 to filters=32 then training works. i was also able to delete this layer entirely and get training to work.
Do so i get reasonable predictions for the test images, and when i run valid i get something like ~70% accuracy

some concluding notes: building with cuda, opencv, openmp do not impact the result, building with cudnn enabled is what causes the issue, training on the cpu everything works
I have only tried with cifar data and my own custom data labeled in cifar style. looking at the code it doesnt look like there is a huge difference between how cifar and imagenet styles are loaded but i will try that next if i can.
I dont know why changing the cifar_small.cfg worked. only changing the 2nd from the last conv layer seemed to fix the training issue, changing other layers didnt help

my end goal is to use darkent19 or 53 but training those had the same issue of always predicting the same label, so i am trying to get cifar to work since training on the gpu is very quick

**Comment by whentheworldfallsapart** [comment]
Update, training with the Alexy Fork of darknet worked, my guess is that some update that Alexy hasnt taken broke training the classifier.

**Comment by whentheworldfallsapart** [comment]
A friend figured it out for me. I forgot to change the cfg file batch and subdivisions from training to test (1 and 1), since that is what i am feeding the model. This only happens with CUDNN on and predict witht he classifier, there is no error output, just the same class predicted no matter what input. The model is trained fine, just predicting with one image and telling darknet that there are multiple images is a problem. This in only a problem for the classifier not the detector. Hopefully this post helps someone

**Comment by lehuiwu** [comment]
The problem is still exist, I have set the batch and subdividions both 1, but the output, just the same class predicted no matter what input.

**Comment by whentheworldfallsapart** [comment]
> The problem is still exist, I have set the batch and subdividions both 1, but the output, just the same class predicted no matter what input.

If you are training on custom data i would drop some print statements in the code and make sure the images and labels are parsed correctly. Also make sure you number of classes is correct. This should not be an issue if the batch ad subdivisions are setup for test, i have been able to trained on custom data as long as i formatted things correct, set the correct number of classes and didnt have extra '_' in the class name

**Comment by tomsang** [comment]
changing the cfg file batch and subdivisions from training to test (1 and 1) still doesn't work when cuda is on.

**Comment by Meteorix** [comment]
@whentheworldfallsapart  thanks,  you save my day!

**Comment by xg590** [comment]
> It seems to be an issue with cudnn in predict, training is actually working just fine, i can train on yolo with gpu and cudnn, opencv etc. Take the trained model and and predict using alexey and get good results. I try to predict in this branch and i get the same class predictions no matter what input i give. If i take that same model and rebuilt darknet without cudnn than the model predicts the correct classification.

I trained on CIFAR with darknet compiled with CUDNN=1 in MAKEFILE. The valid failed miserably. I then recompile the darknet with CUDNN=0. I got pretty good accuracy with the pretrained weights. You saved my day!
